### PR TITLE
implement Worker.terminate() and self.close()

### LINKED ISSUE #4080: Worker postMessage and close race condition
I have stumbled upon a race condition when attempting to close a Worker while a large `Message` is being sent; the following module reproduces the bug 100% of times on my machine:

```typescript
// worker.ts
setTimeout(() => {
  console.log("timeout");
  close();
}, 50);
while (true) {
  await new Promise(done => {
    setTimeout(() => {
      postMessage({ buf: new Uint8Array({ length: 999999 }) });
      done();
    });
  });
}
```

```shell
$ deno --version
deno 0.34.0
v8 8.1.108
typescript 3.7.2

$ RUST_BACKTRACE=1 deno eval "new Worker('./worker.ts',{type:'module'})"
timeout
thread 'deno-worker-USER-WORKER-./worker.ts' panicked at 'Failed to post message to host: SendError { kind: Disconnected }', src/libcore/result.rs:1188:5
stack backtrace:
   0: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
   1: core::fmt::write
   2: std::io::Write::write_fmt
   3: std::panicking::default_hook::{{closure}}
   4: std::panicking::default_hook
   5: std::panicking::rust_panic_with_hook
   6: rust_begin_unwind
   7: core::panicking::panic_fmt
   8: core::result::unwrap_failed
   9: deno::ops::web_worker::op_worker_post_message
  10: deno::state::State::core_op::{{closure}}
  11: deno_core::ops::OpRegistry::call
  12: deno_core::isolate::Isolate::dispatch_op
  13: deno_core::bindings::send
  14: <extern "C" fn(A0) .> R as rusty_v8::support::CFnFrom<F>>::mapping::c_fn
  15: _ZN2v88internal25FunctionCallbackArguments4CallENS0_15CallHandlerInfoE
  16: _ZN2v88internal12_GLOBAL__N_119HandleApiCallHelperILb0EEENS0_11MaybeHandleINS0_6ObjectEEEPNS0_7IsolateENS0_6HandleINS0_10HeapObjectEEESA_NS8_INS0_20FunctionTemplateInfoEEENS8_IS4_EENS0_16BuiltinArgumentsE
  17: _ZN2v88internalL26Builtin_Impl_HandleApiCallENS0_16BuiltinArgumentsEPNS0_7IsolateE
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
fatal runtime error: failed to initiate panic, error 5
[1]    25457 abort      RUST_BACKTRACE=1 /usr/bin/deno eval 
```

### LINKED ISSUE #4342: Worker.prototype.terminate panicked
I'm trying to write a worker runner that run both in browser and deno. I terminate the worker with:
```javascript
if( worker.terminate && typeof worker.terminate === 'function' )
	worker.terminate();
```
That's run well on deno 0.30 and before. But since 0.32, there is a counterfeit method here,  `worker.terminate && typeof worker.terminate === 'function'` returns true now, but if you really to run it, it'll crash. 
On 0.32, a Error with 'Not yet implemented' thrown, we still can use try catch in program. But since 0.33, that change to a panic:
```bash
thread 'deno-worker-foo' panicked at 'not yet implemented', cli/worker.rs:50:5
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:84
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:61
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1025
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1426
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:65
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:50
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:193
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:210
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:471
  11: std::panicking::begin_panic
  12: deno::ops::worker_host::op_host_terminate_worker
  13: deno::state::State::core_op::{{closure}}
  14: deno_core::ops::OpRegistry::call
  15: deno_core::isolate::Isolate::dispatch_op
  16: deno_core::bindings::send
  17: <extern "C" fn(A0) .> R as rusty_v8::support::CFnFrom<F>>::mapping::c_fn
  18: _ZN2v88internal25FunctionCallbackArguments4CallENS0_15CallHandlerInfoE
  19: _ZN2v88internal12_GLOBAL__N_119HandleApiCallHelperILb0EEENS0_11MaybeHandleINS0_6ObjectEEEPNS0_7IsolateENS0_6HandleINS0_10HeapObjectEEESA_NS8_INS0_20FunctionTemplateInfoEEENS8_IS4_EENS0_16BuiltinArgumentsE
  20: _ZN2v88internalL26Builtin_Impl_HandleApiCallENS0_16BuiltinArgumentsEPNS0_7IsolateE
  21: Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_BuiltinExit
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
fatal runtime error: failed to initiate panic, error 5
```

Now we cannot check whether we can this method trustworthy. 
* is there a function here? yep!
* can we run it? nope!
* can we run it in `try{}` and catch something? nope!
* can we check the deno version via globalThis.Deno and not run before it implemented? no way! it's removed.

so we have to write code only for several version of deno.

#### feasible solutions
* remove this method until it's implemented.
* throw Error like 0.32 instead of panic, so we can catch it.
* let globalThis.Deno back so we can check the version of deno.

### PR DESCRIPTION
**Pull request by bartlomieju**
Fixed #4342 
Fixed #4080

### CODE REVIEW

File: cli/tests/subdir/racy_worker.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  }, 50);
+  while (true) {
+    await new Promise((done) => {
+      setTimeout(() => {
+        postMessage({ buf: new Array(999999) });
+        done();
+      });
+    });
+  }
+})();
```
Comment (ry) [inline cli/tests/subdir/racy_worker.js:20]: Nice test. Use top-level await?

File: cli/tests/workers_test.ts [hunk 2]
```diff
@@ -29,10 +29,9 @@ export function createResolvable<T>(): Resolvable<T> {
 
 Deno.test({
   name: "workersBasic",
```
Comment (ry) [inline cli/tests/workers_test.ts:31]: Maybe change the name to "worker terminate" ?

File: cli/tests/subdir/racy_worker.js (hunk 1 above)
Comment (bartlomieju) [reply cli/tests/subdir/racy_worker.js:20]: Copy-pasted from the issue so it can be run in browsers as well

File: cli/web_worker.rs [hunk 3]
```diff
@@ -3,17 +3,59 @@ use crate::ops;
...
 use std::task::Context;
 use std::task::Poll;
+
+#[derive(Clone)]
+pub struct WebWorkerHandle {
+  worker_handle: WorkerHandle,
+  terminate_tx: mpsc::Sender<()>,
+  terminated: Arc<AtomicBool>,
+  isolate_handle: v8::IsolateHandle,
+}
```
Comment (ry) [inline cli/web_worker.rs:29]: Seems reasonable to me.

Consider adding a triple slash comment here describing who accesses this. I think this is a struct that shared between the worker thread itself and the parent worker - both of them may terminate the worker.

File: cli/web_worker.rs [hunk 4]
```diff
@@ -3,17 +3,59 @@ use crate::ops;
...
+  pub fn terminate(&self) {
+    let already_terminated = self.terminated.swap(true, Ordering::Relaxed);
+
+    if !already_terminated {
+      self.isolate_handle.terminate_execution();
+      let mut sender = self.terminate_tx.clone();
+      sender
+        .try_send(())
+        .map_err(ErrBox::from)
+        .expect("Failed to terminate");
```
Comment (ry) [inline cli/web_worker.rs:54]: `.map_err(ErrBox::from)` is unnecessary.

Comment (ry) [reply cli/web_worker.rs:54]: Consider adding some comments here describing the situation in some detail. We are okay with the send failing, correct?

File: cli/web_worker.rs [hunk 5]
```diff
@@ -23,17 +65,48 @@ use std::task::Poll;
 /// `WebWorker`.
 pub struct WebWorker {
   worker: Worker,
-  is_ready: bool,
+  event_loop_idle: bool,
+
+  pub terminated: Arc<AtomicBool>,
```
Comment (ry) [inline cli/web_worker.rs:70]: Is this shared with `WebWorkerHandle`? Wouldn't it be better to just have a reference to the `WebWorkerHandle` ?

File: cli/web_worker.rs [hunk 6]
```diff
@@ -3,17 +3,68 @@ use crate::ops;
...
 use std::task::Context;
 use std::task::Poll;
+
+/// Wrapper for `WorkerHandle` that adds functionality
+/// for terminating workers.
+///
+/// This struct is used by host as well as worker itself.
+///
+/// Host uses it to communicate with worker and terminate it,
+/// while worker uses it only to finish execution on `self.close()`.
```
Comment (ry) [inline cli/web_worker.rs:29]: Thanks

**Pull request merged**